hub/src/server/authentication.js

```diff
@@ -0,0 +1,191 @@
+/* @flow */
+
+import bitcoin from 'bitcoinjs-lib'
+import crypto from 'crypto'
+import { decodeToken, TokenSigner, TokenVerifier } from 'jsontokens'
+import { ecPairToHexString } from 'blockstack'
+import { ValidationError } from './errors'
+import logger from 'winston'
+
+const DEFAULT_STORAGE_URL = 'storage.blockstack.org'
+export const LATEST_AUTH_VERSION = 'v1'
+
+function pubkeyHexToECPair (pubkeyHex) {
+  const pkBuff = Buffer.from(pubkeyHex, 'hex')
+  return bitcoin.ECPair.fromPublicKeyBuffer(pkBuff)
+}
+
+export class V1Authentication {
+  constructor(token: string) {
+    this.token = token
+  }
+
+  static fromAuthPart(authPart: string) {
+    if (!authPart.startsWith('v1:')) {
+      throw new ValidationError('Authorization header should start with v1:')
+    }
+    const token = authPart.slice('v1:'.length)
+    const decodedToken = decodeToken(token)
+    const publicKey = decodedToken.payload.iss
+    if (!publicKey || !decodedToken) {
+      throw new ValidationError('Auth token should be a JWT with at least an `iss` claim')
+    }
+    return new V1Authentication(token)
+  }
+
+  static makeAuthPart(secretKey: bitcoin.ECPair, challengeText: string) {
+    const FOUR_MONTH_SECONDS = 60 * 60 * 24 * 31 * 4
+    const publicKeyHex = secretKey.getPublicKeyBuffer().toString('hex')
+    const salt = crypto.randomBytes(16).toString('hex')
+    const payload = { gaiaChallenge: challengeText,
+                      iss: publicKeyHex,
+                      exp: FOUR_MONTH_SECONDS + (new Date()/1000),
+                      salt }
+    const signerKeyHex = ecPairToHexString(secretKey).slice(0, 64)
+    const token = new TokenSigner('ES256K', signerKeyHex).sign(payload)
+    return `v1:${token}`
+  }
+
+  isAuthenticationValid(address: string, challengeText: string,
+                        options?: { throwOnFailure?: boolean }) {
+    const defaults = {
+      throwOnFailure: true
+    }
+    options = Object.assign({}, defaults, options)
+
+    const decodedToken = decodeToken(this.token)
+    const publicKey = decodedToken.payload.iss
+    const gaiaChallenge = decodedToken.payload.gaiaChallenge
+
+    try {
+      if (! publicKey) {
+        throw new ValidationError('Must provide `iss` claim in JWT.')
+      }
+
+      const issuerAddress = pubkeyHexToECPair(publicKey).getAddress()
+
+      if (issuerAddress !== address) {
+        throw new ValidationError('Address not allowed to write on this path')
+      }
+
+      const verified = new TokenVerifier('ES256K', publicKey).verify(this.token)
+      if (!verified) {
+        throw new ValidationError('Failed to verify supplied authentication JWT')
+      }
+
+      if (gaiaChallenge !== challengeText) {
```

Comment:
I think this check should happen earlier than the signature verification, so we don't put ourselves in a position where the attacker can observe a victim decrypt invalid data.

Reply:
Not sure what the leak would be here -- nothing is being decrypted in the authentication process.

Reply:
Yup.  Updated comment.